guardrails/validators.py

```diff
@@ -1874,3 +1875,286 @@ def query_vector_collection(
 
     def to_prompt(self, with_keywords: bool = True) -> str:
         return ""
+    
+@register_validator(name="provenance-v1", data_type="string")
+class ProvenanceV1(Validator):
+    """Validates that the LLM-generated text is supported by the provided contexts, by prompting another LLM.
+    'LLM-ception'!
+    
+    In order to use this validator, you must provide either:
+    1. a 'query_function' in the metadata. That function should take a string as input (the LLM-generated text) and return a list of relevant 
+    chunks. The list should be sorted in ascending order by the distance between the chunk and the LLM-generated text.
+
+    Example:
+        >>> def query_function(text: str, k: int) -> List[str]:
+        ...     return ["This is a chunk", "This is another chunk"]
+
+        >>> guard = Guard.from_rail(...)
+        >>> guard(
+        ...     openai.ChatCompletion.create(...),
+        ...     prompt_params={...},
+        ...     temperature=0.0,
+        ...     metadata={"query_function": query_function},
+        ... )
+
+    OR
+
+    2. `sources` with an `embed_function` in the metadata. The embed_function should take a string or a list of strings as input and return a np array of floats.
+    The vector should be normalized to unit length.
+
+    Example:
+        ```py
+        def embed_function(text: Union[str, List[str]]) -> np.ndarray:
+            return np.array([[0.1, 0.2, 0.3]])
+
+        guard = Guard.from_rail(...)
+        guard(
+            openai.ChatCompletion.create(...),
+            prompt_params={...},
+            temperature=0.0,
+            metadata={
+                "sources": ["This is a source text"],
+                "embed_function": embed_function
+            },
+        )
+    """
+
+    def __init__(
+        self,
+        validation_method: str = "sentence",
+        openai_model_name: str = "gpt-3.5-turbo",
```

Comment:
You should use an LLM callable for this instead -- you can take a look t the llm_providers.py file.

Reply:
`llm_callable: Optional[str, Callable] = "gpt-3.5-turbo"`

Create a `set_callable` method
- runs at validator initialization and sets a `self._llm_callable`
- if callable is string, then you do the switch case thing
- if callable doesn't fall inside switch case, then raise an error
- if callable is of type `Callable`, then just call the callable directly